### PR TITLE
Support Long/Float dimensions with generic dimension handling interface

### PR DESCRIPTION
**Pull request by jon-wei**
This PR adds support for Long and Float numeric dimensions using a new dimension handling interface.

It depends on the following two PRs:

https://github.com/druid-io/druid-api/pull/75
https://github.com/druid-io/druid/pull/2607

Implementation notes:
* At a high level, type-specific handling for dimensions in the IncrementalIndex, IndexMerger, and querying paths has been extracted into a new interface. There are 4 associated classes with the new interface:
 - DimensionHandler: Top level class, provides comparators and used to create the other 3 classes
 - DimensionIndexer: Handles logic related to in-memory ingestion of rows (IncrementalIndex)
 - DimensionMerger: Handles logic related to merging of segments (IndexMerger)
 - DimensionColumnReader: Handles logic related to reading from QueryableIndex/Columns 
* There are String/Long/Float implementations of these classes.

* Dictionary encoding is not used for Long and Float dimensions. No index structures are currently created for numeric dimensions. Numeric dims will use the existing long/float column formats on disk.
* The DimensionSelector interface now has additional functions for retrieving rows from numeric dimension columns. There are separate functions for long/float primitives, to avoid primitive boxing/unboxing. Benchmark runs showed a fairly significant increase in processing time when using a Comparable instead of IndexedInts directly.
* As there are no bitmap indexes on numeric dims, a full scan will be performed on numeric dims for filtering logic that uses filter.getBitmapIndex()
* To support the requirement above, ColumnSelectorBitmapIndexSelector now supports bitmap generation from the result of a column scan
* SelectorFilter can do direct numeric comparisons on Long/Float columns, no string conversion needed
* Long and Float values will be converted to Strings for any filters or extraction functions that require String inputs.
* null values for Long and Float dims will be converted to 0L or 0.0F, respectively.
* Three benchmarks have been added:
 - IncrementalIndexAddRowsBenchmark: adds rows with different value types to IncrementalIndex
 - IndexedIntsWrapping: evaluates performance of wrapping/unboxing of ints
 - QueryableIndexLoadingBenchmark: evaluates performance of reading from a QueryableIndex with/without integer boxing, using DimensionSelectors

Unit test notes:

* The druid.sample.tsv file has been modified to include Long and Float dimension columns (market_long, market_float, quality_long, quality_float). The original druid.sample.tsv file has been kept in the repo.
* The numeric values for these new dimensions have a 1-to-1 mapping with the values in 'market' and 'quality'
* For larger test suites like GroupByQueryRunnerTest and TopN, I have parameterized the use of the "quality" and "market" dimensions; in different test iterations, the tests will use the long or float versions of these dimension columns.
* Smaller test suites have separate functions for testing numeric dimensions.

Example benchmark run showing difference between using Comparable (loadQIndexWrap) and IndexedInts (loadQIndex)
```
64 dim, 20000 cardinality
Benchmark                                       Mode  Cnt  Score   Error  Units
QueryableIndexLoadingBenchmark.loadQIndexWrap  avgt  100  1.414 ± 0.016  us/op
QueryableIndexLoadingBenchmark.loadQIndex       avgt  100  1.127 ± 0.009  us/op

Benchmark                                       Mode  Cnt  Score   Error  Units
QueryableIndexLoadingBenchmark.loadQIndexWrap  avgt  100  1.405 ± 0.008  us/op
QueryableIndexLoadingBenchmark.loadQIndex       avgt  100  1.166 ± 0.011  us/op
```

### CODE REVIEW

File: benchmarks/src/main/java/io/druid/benchmark/IndexedIntsWrappingBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,382 @@
+package io.druid.benchmark;
```
Comment (fjy) [inline benchmarks/src/main/java/io/druid/benchmark/IndexedIntsWrappingBenchmark.java]: header

File: benchmarks/src/main/java/io/druid/benchmark/QueryableIndexLoadingBenchmark.java [hunk 2]
```diff
@@ -0,0 +1,293 @@
+package io.druid.benchmark;
```
Comment (fjy) [inline benchmarks/src/main/java/io/druid/benchmark/QueryableIndexLoadingBenchmark.java]: header here and other files that are missing it

File: docs/content/development/geo.md [hunk 3]
```diff
@@ -5,7 +5,7 @@ layout: doc_page
 Druid supports filtering specially spatially indexed columns based on an origin and a bound.
 
 # Spatial Indexing
-In any of the data specs, there is the option of providing spatial dimensions. For example, for a JSON data spec, spatial dimensions can be specified as follows:
+@Deprecated. In any of the data specs, there is the option of providing spatial dimensions. For example, for a JSON data spec, spatial dimensions can be specified as follows:
```
Comment (fjy) [inline docs/content/development/geo.md]: can we highlight Deprecated with a note-caution tag?

File: processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java [hunk 4]
```diff
@@ -42,21 +44,30 @@
...
+import io.druid.segment.column.ValueType;
+import io.druid.segment.data.IndexedFloats;
 import io.druid.segment.data.IndexedInts;
+import io.druid.segment.data.IndexedLongs;
 import io.druid.segment.filter.Filters;
 import org.joda.time.DateTime;
 import org.joda.time.Interval;
 
 import javax.annotation.Nullable;
+import java.awt.*;
```
Comment (fjy) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java]: can we include necessary imports instead of including *?

File: processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java [hunk 5]
```diff
@@ -143,25 +156,85 @@ public void close() throws IOException
     );
   }
 
+  private static void getBytesFromBuffer(ByteBuffer src, byte[] dst, int srcOffset, int readLen) {
+    for(int i = 0; i < readLen; i++) {
```
Comment (fjy) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java]: there's some formatting problems here

File: benchmarks/src/main/java/io/druid/benchmark/QueryableIndexLoadingBenchmark.java [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+  {
+    if(initialized) {
+      return;
+    }
+
+    rng = new Random(9999);
+    rng2 = new Random(9999);
+
+
+    /*
```
Comment (fjy) [inline benchmarks/src/main/java/io/druid/benchmark/QueryableIndexLoadingBenchmark.java]: can we remove commented out code?

File: processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java (hunk 5 above)
Comment (fjy) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java]: also, can we define static methods at the top of files?

File: processing/src/main/java/io/druid/segment/FloatDimensionHandler.java [hunk 7]
```diff
@@ -0,0 +1,518 @@
...
+          dimensionValues = new Comparable[dimValuesList.size()];
+          for (int i = 0; i < dimValuesList.size(); i++) {
+            transformedVal = FLOAT_TRANSFORMER.apply(dimValuesList.get(i));
+            dimensionValues[i] = transformedVal;
+          }
+        } else {
+          transformedVal = FLOAT_TRANSFORMER.apply(dimValues);
+          dimensionValues = new Comparable[]{transformedVal};
+        }
+        //minValue = Float.compare(transformedVal, minValue) < 0 ? transformedVal : minValue;
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/FloatDimensionHandler.java]: remove dead code

File: processing/src/main/java/io/druid/segment/FloatDimensionHandler.java [hunk 8]
```diff
@@ -0,0 +1,518 @@
...
+    private float maxValue = Float.MIN_VALUE;
+
+    @Override
+    public Comparable[] processRowValsToIndexKey(Object dimValues)
+    {
+      final Comparable[] dimensionValues;
+      Float transformedVal = 0.0f;
+      try {
+        if (dimValues == null) {
+          //return new Comparable []{};
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/FloatDimensionHandler.java]: remove commented out code here and in other places

File: processing/src/main/java/io/druid/segment/FloatDimensionHandler.java [hunk 9]
```diff
@@ -0,0 +1,518 @@
...
+    @Override
+    public ImmutableBitmap getBitmapIndex(
+        String value, BitmapFactory bitmapFactory, int numRows
+    )
+    {
+      if (column == null) {
+        return bitmapFactory.makeEmptyImmutableBitmap();
+      }
+      final float parsedFloat = Float.parseFloat(value);
+      //final float parsedFloat = value;
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/FloatDimensionHandler.java]: remove

File: processing/src/main/java/io/druid/segment/FloatDimensionHandler.java [hunk 10]
```diff
@@ -0,0 +1,518 @@
...
+
+    @Override
+    public ImmutableBitmap getBitmapIndex(
+        String value, BitmapFactory bitmapFactory, int numRows
+    )
+    {
+      if (column == null) {
+        return bitmapFactory.makeEmptyImmutableBitmap();
+      }
+      final float parsedFloat = Float.parseFloat(value);
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/FloatDimensionHandler.java]: what happens if value isn't a float?

File: processing/src/main/java/io/druid/segment/FloatDimensionSelector.java [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+  public IndexedFloats getFloatRow()
+  {
+    final float val = selector.get();
+    return new IndexedFloats()
+    {
+      @Override
+      public int size()
+      {
+        // only single-value numerics are supported
+        return 1;
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/FloatDimensionSelector.java]: where does the code check for this constraint?

File: processing/src/main/java/io/druid/segment/FloatDimensionSelector.java [hunk 12]
```diff
@@ -0,0 +1,130 @@
+package io.druid.segment;
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/FloatDimensionSelector.java]: header

File: processing/src/main/java/io/druid/segment/FloatMetricColumnSerializer.java [hunk 13]
```diff
@@ -70,6 +70,12 @@ public void serialize(Object obj) throws IOException
   public void close() throws IOException
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/FloatMetricColumnSerializer.java]: can we call this method something else?

File: processing/src/main/java/io/druid/segment/FloatUnencodedDimensionSelector.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
+package io.druid.segment;
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/FloatUnencodedDimensionSelector.java]: is this file needed?

File: processing/src/main/java/io/druid/segment/IndexMerger.java [hunk 15]
```diff
@@ -862,107 +791,33 @@ public Metadata apply(IndexableAdapter input)
...
-        for (Integer row : CombiningIterable.createSplatted(
-            convertedInverteds,
-            Ordering.<Integer>natural().nullsFirst()
-        )) {
-          if (row != INVALID_ROW) {
-            bitset.add(row);
-          }
-        }
+    for (DimensionMerger merger : mergers) {
+      merger.closeWritersV8(rowNumConversions, out, spatialOut);
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/IndexMerger.java]: do we still want to support v8? can we just use v9?

File: processing/src/main/java/io/druid/segment/IndexMerger.java [hunk 16]
```diff
@@ -1007,18 +862,41 @@ public Metadata apply(IndexableAdapter input)
...
+      }
+
+      if (capabilities.getType() == ValueType.LONG && !capabilities.isDictionaryEncoded()) {
+        handlers[i] = new LongDimensionHandler(dimName);
+      }
+
+      if (capabilities.getType() == ValueType.FLOAT && !capabilities.isDictionaryEncoded()) {
+        handlers[i] = new FloatDimensionHandler(dimName);
+      }
+    }
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/IndexMerger.java]: what happens if the cases are not met? should there be errors?

File: processing/src/main/java/io/druid/segment/LongDimensionHandler.java [hunk 17]
```diff
@@ -0,0 +1,513 @@
...
+    private long minValue = Long.MAX_VALUE;
+    private long maxValue = Long.MIN_VALUE;
+    @Override
+    public Comparable[] processRowValsToIndexKey(Object dimValues)
+    {
+      final Comparable[] dimensionValues;
+      Long transformedVal = 0L;
+      try {
+        if (dimValues == null) {
+          //return new Comparable []{};
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/LongDimensionHandler.java]: remove

File: processing/src/main/java/io/druid/segment/LongDimensionHandler.java [hunk 18]
```diff
@@ -0,0 +1,513 @@
...
+          for (int i = 0; i < dimValuesList.size(); i++) {
+            transformedVal = LONG_TRANSFORMER.apply(dimValuesList.get(i));
+            dimensionValues[i] = transformedVal;
+          }
+        } else {
+          transformedVal = LONG_TRANSFORMER.apply(dimValues);
+          dimensionValues = new Comparable[]{transformedVal};
+        }
+        //minValue = Long.compare(transformedVal, minValue) < 0 ? transformedVal : minValue;
+        //maxValue = Long.compare(transformedVal, maxValue) > 0 ? transformedVal : maxValue;
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/LongDimensionHandler.java]: remove

File: processing/src/main/java/io/druid/segment/LongDimensionSelector.java [hunk 19]
```diff
@@ -0,0 +1,145 @@
+package io.druid.segment;
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/LongDimensionSelector.java]: header

File: processing/src/main/java/io/druid/segment/NullDimensionSelector.java [hunk 20]
```diff
@@ -81,4 +101,136 @@ public int lookupId(String name)
...
+  public Comparable getExtractedValueLong(long val)
+  {
+    return val;
+  }
+
+  @Override
+  public IndexedFloats getFloatRow()
+  {
+    throw new UnsupportedOperationException("AAAAA");
+    /*
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/NullDimensionSelector.java]: remove

File: processing/src/main/java/io/druid/segment/NullDimensionSelector.java [hunk 21]
```diff
@@ -81,4 +101,136 @@ public int lookupId(String name)
...
+  @Override
+  public Comparable getExtractedValueLong(long val)
+  {
+    return val;
+  }
+
+  @Override
+  public IndexedFloats getFloatRow()
+  {
+    throw new UnsupportedOperationException("AAAAA");
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/NullDimensionSelector.java]: better error msg :P

File: processing/src/main/java/io/druid/segment/NullDimensionSelector.java [hunk 22]
```diff
@@ -81,4 +101,136 @@ public int lookupId(String name)
...
+    }
+    return rowVals;
+  }
+
+  @Override
+  public Comparable getExtractedValueFromUnencoded(Comparable rowVal)
+  {
+    return rowVal;
+  }
+  */
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/NullDimensionSelector.java]: remove

File: processing/src/main/java/io/druid/segment/QueryableIndexIndexableAdapter.java [hunk 23]
```diff
@@ -247,22 +217,35 @@ public Rowboat next()
...
+                for (int j = 0; j < theVals.length; ++j) {
+                  theVals[j] = dimVals.get(j);
+                }
 
-              dims[dimIndex++] = theVals;
+                dims[dimIndex++] = theVals;
+              } else if (column instanceof GenericColumn) {
+                GenericColumn genericCol = (GenericColumn) column;
+                switch(genericCol.getType()) {
+                  case LONG:
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/QueryableIndexIndexableAdapter.java]: as i read through the code, im a little worried of all the places that require switch statements such as this. It'll be easy to miss cases and cause bugs as we need new types. I wonder if we can commonalize this logic somehow

File: processing/src/main/java/io/druid/segment/SimpleQueryableIndex.java [hunk 24]
```diff
@@ -108,4 +117,46 @@ public Metadata getMetadata()
...
+        handler = new StringDimensionHandler(dim);
+        dimensionHandlers.put(dim, handler);
+      }
+
+      if (capabilities.getType() == ValueType.LONG && !capabilities.isDictionaryEncoded()) {
+        handler = new LongDimensionHandler(dim);
+        dimensionHandlers.put(dim, handler);
+      }
+
+      if (capabilities.getType() == ValueType.FLOAT && !capabilities.isDictionaryEncoded()) {
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/SimpleQueryableIndex.java]: this logic seems C&Ped from code I saw before

File: processing/src/main/java/io/druid/query/select/SelectQueryEngine.java [hunk 25]
```diff
@@ -121,22 +126,44 @@
...
-                    theEvent.put(dim, dimVal);
-                  } else {
-                    List<String> dimVals = Lists.newArrayList();
-                    for (int i = 0; i < vals.size(); ++i) {
-                      dimVals.add(selector.lookupName(vals.get(i)));
-                    }
-                    theEvent.put(dim, dimVals);
+                  Object dimVals;
+                  ValueType type = capabilities == null ? ValueType.STRING : capabilities.getType();
+                  //switch(capabilities.getType()) {
```
Comment (fjy) [inline processing/src/main/java/io/druid/query/select/SelectQueryEngine.java]: remove

File: processing/src/main/java/io/druid/query/topn/DimExtractionTopNAlgorithm.java [hunk 26]
```diff
@@ -90,36 +99,66 @@ public TopNParams makeInitParams(
...
+            aggregator.aggregate();
+          }
+        }
+        cursor.advance();
+      }
+    } else {
+      while (!cursor.isDone()) {
+        // Use the aggregatesStore directly since there is no encoding
+        Comparable key;
+        switch(columnCapabilities.getType()) {
```
Comment (fjy) [inline processing/src/main/java/io/druid/query/topn/DimExtractionTopNAlgorithm.java]: can we move this logic outside the cursor loop? should help to improve performance

File: processing/src/main/java/io/druid/query/topn/DimExtractionTopNAlgorithm.java [hunk 27]
```diff
@@ -90,36 +99,66 @@ public TopNParams makeInitParams(
...
-      final IndexedInts dimValues = dimSelector.getRow();
-
-      for (int i = 0; i < dimValues.size(); ++i) {
-
-        final int dimIndex = dimValues.get(i);
-        Aggregator[] theAggregators = rowSelector[dimIndex];
-        if (theAggregators == null) {
-          final String key = dimSelector.lookupName(dimIndex);
-          theAggregators = aggregatesStore.get(key);
+    if (columnCapabilities.isDictionaryEncoded()) {
```
Comment (fjy) [inline processing/src/main/java/io/druid/query/topn/DimExtractionTopNAlgorithm.java]: it'd be really nice to not have an if/else here especially later on as we have more conditions

File: processing/src/main/java/io/druid/segment/ComparableDimensionSelector.java [hunk 28]
```diff
@@ -0,0 +1,102 @@
+package io.druid.segment;
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/ComparableDimensionSelector.java]: header

**Comment by jon-wei** [comment]
@fjy added headers and removed dead code

File: processing/src/main/java/io/druid/segment/FloatDimensionSelector.java (hunk 11 above)
Comment (jon-wei) [inline processing/src/main/java/io/druid/segment/FloatDimensionSelector.java]: @fjy Added a size() > 1 check to LongDimensionIndexer and FloatDimensionIndexer, they'll throw an exception now if a multi-value row is passed to them

File: docs/content/development/geo.md (hunk 3 above)
Comment (jon-wei) [inline docs/content/development/geo.md]: @fjy added note-caution tag around deprecation note

File: processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java (hunk 4 above)
Comment (jon-wei) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java]: @fjy that was a junk import (probably from auto-complete), i've deleted it

File: processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java (hunk 5 above)
Comment (jon-wei) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java]: @fjy moved methods and adjusted formatting

File: processing/src/main/java/io/druid/segment/FloatDimensionHandler.java (hunk 7 above)
Comment (jon-wei) [inline processing/src/main/java/io/druid/segment/FloatDimensionHandler.java]: @fjy Removed min/max calculation for Long/Floats, added a note about that to DimensionIndexer interface

File: processing/src/main/java/io/druid/segment/FloatDimensionHandler.java (hunk 10 above)
Comment (jon-wei) [inline processing/src/main/java/io/druid/segment/FloatDimensionHandler.java]: @fjy Added a try/catch for NumberFormatException, will throw a ParseException if that occurs

File: processing/src/main/java/io/druid/segment/FloatMetricColumnSerializer.java (hunk 13 above)
Comment (jon-wei) [inline processing/src/main/java/io/druid/segment/FloatMetricColumnSerializer.java]: @fjy Renamed the close() method with a File parameter to closeFile()

File: processing/src/main/java/io/druid/segment/FloatUnencodedDimensionSelector.java (hunk 14 above)
Comment (jon-wei) [inline processing/src/main/java/io/druid/segment/FloatUnencodedDimensionSelector.java]: @fjy deleted unnecessary file

File: processing/src/main/java/io/druid/segment/IndexMerger.java (hunk 16 above)
Comment (jon-wei) [inline processing/src/main/java/io/druid/segment/IndexMerger.java]: @fjy I've moved this capabilities -> DimensionHandler code to a static method on a util class, DimensionHandlerUtil.

It will throw an IAE if capabilities type is not String/Long/Float or if the dictionary encoding/bitmap indexing settings are incorrect for a given type

File: processing/src/main/java/io/druid/segment/SimpleQueryableIndex.java (hunk 24 above)
Comment (jon-wei) [inline processing/src/main/java/io/druid/segment/SimpleQueryableIndex.java]: @fjy Moved this code to DimensionHandlerUtil

File: processing/src/main/java/io/druid/query/topn/DimExtractionTopNAlgorithm.java (hunk 26 above)
Comment (jon-wei) [inline processing/src/main/java/io/druid/query/topn/DimExtractionTopNAlgorithm.java]: @fjy Cursor iteration is now inside the type-specific switch cases

File: processing/src/main/java/io/druid/query/topn/DimExtractionTopNAlgorithm.java (hunk 27 above)
Comment (jon-wei) [inline processing/src/main/java/io/druid/query/topn/DimExtractionTopNAlgorithm.java]: @fjy Moved the "dictionary encoding == true" block to a String switch case

**Comment by fjy** [comment]
@jon-wei merge conflicts

File: benchmarks/src/main/java/io/druid/benchmark/QueryableIndexLoadingBenchmark.java (hunk 6 above)
Comment (jon-wei) [inline benchmarks/src/main/java/io/druid/benchmark/QueryableIndexLoadingBenchmark.java]: @fjy 

It's sort of rough, but I had two "modes" of behavior at that area in the benchmark, one was to generate a fresh QueryableIndex, and the other was to load a pre-generated index from disk.

I've uncommented out that block which handles the former case, left a commented line and comment about loading from a file

File: processing/src/main/java/io/druid/segment/NullDimensionSelector.java (hunk 21 above)
Comment (jon-wei) [inline processing/src/main/java/io/druid/segment/NullDimensionSelector.java]: @fjy I had the exception there for testing/debugging, I've restored the commented out block

**Comment by jon-wei** [comment]
@fjy I'll do a rebase after https://github.com/druid-io/druid/pull/2607 is merged, so I can get rid of that as a dependent commit

**Comment by jon-wei** [comment]
Rebased

**Comment by jon-wei** [comment]
I will try to reduce the size of this PR by splitting out the generic dimension interface changes, with support for only String dimensions.

Long/Float support will then be a separate follow-on PR that uses the generic dimension interface.

**Comment by jon-wei** [comment]
Closing this PR, I've moved the new dimension interface into a separate PR:
https://github.com/druid-io/druid/pull/2760
